src/backends/policy_engine/souffle/check_policy_compliance.cc

```diff
@@ -61,6 +63,51 @@ absl::StatusOr<std::string> ReadFileContents(std::filesystem::path file_path) {
   return string_stream.str();
 }
 
+enum class ReturnCode : uint8_t { PASS = 0, FAIL, ERROR };
+
+int UnwrapExitCode(ReturnCode code) { return static_cast<int>(code); }
+
+struct IrGraphComponents {
+  std::unique_ptr<raksha::ir::Module> ir_module;
+  std::unique_ptr<raksha::ir::SsaNames> ir_ssa_names;
+  std::unique_ptr<raksha::ir::IRContext> ir_context;
+};
+
+absl::StatusOr<IrGraphComponents> GetIrGraphComponentsFromIrPath(
+    absl::string_view ir_path) {
+  raksha::ir::IrProgramParser ir_parser;
+  absl::StatusOr<std::string> ir_string = ReadFileContents(ir_path);
+  if (ir_string.ok()) {
+    auto [context, module, ssa_names] = ir_parser.ParseProgram(*ir_string);
+    return IrGraphComponents{.ir_module = std::move(module),
+                             .ir_ssa_names = std::move(ssa_names),
+                             .ir_context = std::move(context)};
+  } else {
+    return ir_string.status();
+  }
+}
+
+absl::StatusOr<IrGraphComponents> GetIrGraphComponentsFromProtoPath(
+    absl::string_view proto_path) {
+  absl::StatusOr<std::string> proto_string = ReadFileContents(proto_path);
+  if (proto_string.ok()) {
+    std::unique_ptr<raksha::ir::IRContext> context =
+        std::make_unique<raksha::ir::IRContext>(); /*...*/
+    auto ir_result = raksha::ir::ProtoToIR::Convert(*context, *proto_string);
+    if (ir_result.ok()) {
+      return IrGraphComponents{
+          .ir_module = std::move(ir_result.value().module),
+          .ir_ssa_names = std::move(ir_result.value().ssa_names),
+          .ir_context = std::move(context)};
+    } else {
+      return ir_result.status();
+    }
+  } else {
+    LOG(ERROR) << "Error reading proto file: " << proto_string.status();
+    return proto_string.status();
+  }
```

Comment:
What I see here is some code that basically wants to turn one `StatusOr` into another concisely. [Rust provides a `map` method for this purpose](https://doc.rust-lang.org/std/result/enum.Result.html#method.map), and there's no reason why we can't do that as well. We have a small little functional-programming-in-C++ library in `//src/common/utils`; perhaps we could add a `statusor_map` function there for handling this case more concisely.

Reply:
I think that's a great idea, maybe I should do it in a follow up PR? I feel like if I'm adding some useful utility function for everyone it should be a separate commit + tests?

Reply:
Totally, follow on PR is probably best to keep this PR focused on one aspect.